samza-core/src/main/java/org/apache/samza/util/BoundedFifoCache.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.util;
+
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+/**
+ * FIFO cache of size cacheSize
+ *
+ * This class is not thread-safe
+ */
+public class BoundedFifoCache<T> {
```

Comment:
This is really a BoundedLinkedHashSet, the fact that we're using it as a fifo cache is irrelevant.

Reply:
I think the name is intuitive enough, it's a BoundedFifoCache, what it's using underneath should be abstracted. But I am ok with using BoundedLinkedHashSet as well

Reply:
I agree its intuitive for this use-case, but the constituents may not always be "cached" elements. They can for example be a time-window of actions that are held/stored in this data structure when its not being used as a cache. 
This is akin to how a DelayedQueue is called that, even 
though 99% of the time it'd be used to schedule input to a scheduler.
The name of the class is supposed to describe what it does, not how/what it can be used for.